utilities/scale-json/src/lib.rs

```diff
@@ -0,0 +1,391 @@
+					|(all_bytes, mut values), res| {
+						res.and_then(|item| {
+							let decoded_item = item.decode_with_visitor(self)?;
+							let is_byte = item.bytes().len() == 1 && decoded_item.is_u64();
+							values.push(decoded_item);
+							Ok((all_bytes && is_byte, values))
+						})
+					},
+				)?;
+				if is_bytes {
+					Ok(serde_json::Value::String(untrimmed_hex(
+						all_values
+							.into_iter()
+							.map(|json| json.as_u64().unwrap() as u8)
+							.collect::<Vec<u8>>(),
+					)))
+				} else {
+					Ok(serde_json::Value::Array(all_values))
+				}
+			}
+		)+
+	};
+}
+
+impl<R: TypeResolver> scale_decode::Visitor for RawJsonVisitor<R> {
+	type Value<'scale, 'resolver> = serde_json::Value;
+	type Error = DecodeError;
+	type TypeResolver = R;
+
+	decode_to_number! {
+		visit_u8, u8,
+		visit_u16, u16,
+		visit_u32, u32,
+		visit_i8, i8,
+		visit_i16, i16,
+		visit_i32, i32,
+	}
+
+	decode_to_hex! {
+		visit_u64, u64 | to_be_bytes,
+		visit_u128, u128 | to_be_bytes,
+		visit_u256, &'scale [u8; 32],
+		visit_i64, i64 | to_be_bytes,
+		visit_i128, i128 | to_be_bytes,
+		visit_i256, &'scale [u8; 32],
+	}
+
+	decode_to_array! {
+		visit_tuple, Tuple,
+	}
+
+	decode_to_array_or_hex! {
+		visit_sequence, Sequence,
+		visit_array, Array,
+	}
+
+	fn visit_bool<'scale, 'resolver>(
+		self,
+		value: bool,
+		_type_id: scale_decode::visitor::TypeIdFor<Self>,
+	) -> Result<Self::Value<'scale, 'resolver>, Self::Error> {
+		Ok(serde_json::Value::Bool(value))
+	}
+
+	fn visit_char<'scale, 'resolver>(
+		self,
+		value: char,
+		_type_id: scale_decode::visitor::TypeIdFor<Self>,
+	) -> Result<Self::Value<'scale, 'resolver>, Self::Error> {
+		Ok(serde_json::Value::String(value.to_string()))
+	}
+
+	fn visit_str<'scale, 'resolver>(
+		self,
+		value: &mut scale_decode::visitor::types::Str<'scale>,
+		_type_id: scale_decode::visitor::TypeIdFor<Self>,
+	) -> Result<Self::Value<'scale, 'resolver>, Self::Error> {
+		Ok(serde_json::Value::String(value.as_str()?.to_string()))
+	}
+
+	fn visit_bitsequence<'scale, 'resolver>(
+		self,
+		value: &mut scale_decode::visitor::types::BitSequence<'scale>,
+		_type_id: scale_decode::visitor::TypeIdFor<Self>,
+	) -> Result<Self::Value<'scale, 'resolver>, Self::Error> {
+		let s = value
+			.decode()?
+			.map(|r| r.map(|b| if b { '1' } else { '0' }))
+			.collect::<Result<String, _>>()?;
+		Ok(serde_json::Value::String(format!("0b{}", s)))
+	}
+
+	fn visit_composite<'scale, 'resolver>(
+		self,
+		value: &mut scale_decode::visitor::types::Composite<'scale, 'resolver, Self::TypeResolver>,
+		_type_id: scale_decode::visitor::TypeIdFor<Self>,
+	) -> Result<Self::Value<'scale, 'resolver>, Self::Error> {
+		if value.has_unnamed_fields() {
+			if value.remaining() == 1 {
+				// Simple wrapper struct. No need for array.
+				let field = value.next().unwrap()?;
+				assert!(field.name().is_none(), "Checked for unnamed fields.");
+				Ok(field.decode_with_visitor(RawJsonVisitor::<R>::default())?)
```

Comment:
Composite could be a struct or it could be contents of an enum variant (see below, this is called from visit_variant).

A composite with a single unnamed field would be, for example: 

```rust
struct This(..)
```

or:

```rust
enum X {
	A(This),
}
```

In terms of how this is encoded in Scale, an enum is equivalent to struct (just encode the fields). It's the metadata that gives context about how to *decode* it and that is what determines if it calls visit_composite (which has a list of fields that *might* be named) vs visit_tuple (just list of fields with no names) vs visit_array or visit_sequence (lists of unnamed fields that all have the same type). Array and sequence are actually treated the same in scale_decode.